framework/filters/AccessControl.php

```diff
@@ -116,8 +114,15 @@ public function init()
      */
     public function beforeAction($action)
     {
+        if ($this->user !== false) {
+            if (is_string($this->user) && $action->controller->module->has($this->user)) {
+                $this->user = $action->controller->module->get($this->user);
+            }
+            $this->user = Instance::ensure($this->user, User::className());
+        }
+
         $user = $this->user;
-        $request = Yii::$app->getRequest();
+        $request = Yii::get('request');
```

Comment:
Why not `$action->controller->module->get(...)` ?

Reply:
Hmm... still the parameter is an action right?
Alternatively we could check the owner of the behavior? (I've been considering implementing `ServiceLocator` on controllers *-))

Reply:
Such check is already perfromed inside `Yii::get()` what is the sense of repeating it at every filter?

Reply:
The global static getter makes a lot of assumptions based on current structure of the framework; this means that that structure becomes more and more rigid since any change changes these global getters that are potentially used anywhere in a code base.

Reply:
This is exactly what you have pushed framework towards at #14350.
Instead of using explicit predefined methods for retrieval of the components, like [yii\web\Application::getRequest()](https://github.com/yiisoft/yii2/blob/master/framework/web/Application.php#L158) or [yii\web\Application::getSession()](https://github.com/yiisoft/yii2/blob/master/framework/web/Application.php#L176), you suggest usage of following constructions:

```
$controller->module->get('request'); // instead of `Yii::$app->getRequest()`

$controller->module->get('session'); // instead of `Yii::$app->getSession()`
```

as they allows triggering of tree traversal component search.

You have placed restrictions on component structure for the application and module yourself.
Why there are compains on it now?

Reply:
These are not dependent on the application having a single active controller or module; they are just dependent on direct relationships. These chains could (I'm not sure if they should) even be shortened by having a controller act like a service locator as well, but that's a whole different discussion.

I stand by my request and actually use tree traversal in my applications in production, it works and has valid use cases.